File: ts/packages/luzid-grpc-client/src/client/validator.ts
import { createClient } from '@luzid/grpc-connection'
import type { Channel } from '@luzid/grpc-connection'

import {
  ValidatorInfo,
  ValidatorInfoSubClient,
  ValidatorInfoSubDefinition,
  ValidatorOpsRequest,
  ValidatorOpsResponse,
  ValidatorOpsServiceClient,
  ValidatorOpsServiceDefinition,
  ValidatorStats,
  ValidatorStatsSubClient,
  ValidatorStatsSubDefinition,
  ValidatorStatus,
  ValidatorStatusSubClient,
  ValidatorStatusSubDefinition,
} from '@luzid/grpc'

export {
  ValidatorInfo,
  ValidatorOperation,
  ValidatorOpsRequest,
  ValidatorOpsResponse,
  ValidatorStats,
  ValidatorStatus,
  validatorOperationFromJSON,
  validatorOperationToJSON,
} from '@luzid/grpc'

class ValidatorOpsClient {
  private readonly client: ValidatorOpsServiceClient

  constructor(channel: Channel) {
    this.client = createClient(ValidatorOpsServiceDefinition, channel)
  }

  validatorOps(request: ValidatorOpsRequest): Promise<ValidatorOpsResponse> {
    return this.client.validatorOps(request)
  }
}

class _ValidatorInfoSubClient {
  private readonly client: ValidatorInfoSubClient

  constructor(channel: Channel) {
    this.client = createClient(ValidatorInfoSubDefinition, channel)
  }

  subValidatorInfo(opts: {
    emitCurrent: boolean
  }): AsyncIterable<ValidatorInfo> {
    return this.client.subValidatorInfo(opts)
  }
}

class _ValidatorStatusSubClient {
  private readonly client: ValidatorStatusSubClient

  constructor(channel: Channel) {
    this.client = createClient(ValidatorStatusSubDefinition, channel)
  }

  subValidatorStatus(opts: {
    emitCurrent: boolean
  }): AsyncIterable<ValidatorStatus> {
    return this.client.subValidatorStatus(opts)
  }
}

class _ValidatorStatsSubClient {
  private readonly client: ValidatorStatsSubClient

  constructor(channel: Channel) {
    this.client = createClient(ValidatorStatsSubDefinition, channel)
  }

  subValidatorStats(opts: {
    emitCurrent: boolean
  }): AsyncIterable<ValidatorStats> {
    return this.client.subValidatorStats(opts)
  }
}

// -----------------
// Consolidated ValidatorClient
// -----------------
export class ValidatorClient {
  private readonly validatorOpsClient: ValidatorOpsClient
  private readonly validatorInfoSubClient: _ValidatorInfoSubClient
  private readonly validatorStatusSubClient: _ValidatorStatusSubClient
  private readonly validatorStatsSubClient: _ValidatorStatsSubClient

  constructor(channel: Channel) {
    this.validatorOpsClient = new ValidatorOpsClient(channel)
    this.validatorInfoSubClient = new _ValidatorInfoSubClient(channel)
    this.validatorStatusSubClient = new _ValidatorStatusSubClient(channel)
    this.validatorStatsSubClient = new _ValidatorStatsSubClient(channel)
  }

  validatorOps(request: ValidatorOpsRequest): Promise<ValidatorOpsResponse> {
    return this.validatorOpsClient.validatorOps(request)
  }

  subValidatorInfo(opts: {
    emitCurrent: boolean
  }): AsyncIterable<ValidatorInfo> {
    return this.validatorInfoSubClient.subValidatorInfo(opts)
  }

  subValidatorStatus(opts: {
    emitCurrent: boolean
  }): AsyncIterable<ValidatorStatus> {
    return this.validatorStatusSubClient.subValidatorStatus(opts)
  }

  subValidatorStats(opts: {
    emitCurrent: boolean
  }): AsyncIterable<ValidatorStats> {
    return this.validatorStatsSubClient.subValidatorStats(opts)
  }
}

File: dart/luzid-grpc-client/lib/src/client/validator.dart
import 'package:grpc/grpc_connection_interface.dart';
import 'package:luzid_grpc/luzid_grpc.dart';
import 'package:luzid_grpc_client/src/core/channel.dart';

// -----------------
// ValidatorOpsClient
// -----------------
class ValidatorOpsClient {
  final ValidatorOpsServiceClient _client;

  ValidatorOpsClient(ClientChannelBase channel)
      : _client = ValidatorOpsServiceClient(channel);

  Future<ValidatorOpsResponse> validatorOps(ValidatorOpsRequest request) {
    return _client.validatorOps(request);
  }
}

// -----------------
// ValidatorStatusSubClient
// -----------------
class _ValidatorStatusSubClient {
  final ValidatorStatusSubClient _client;

  _ValidatorStatusSubClient(ClientChannelBase channel)
      : _client = ValidatorStatusSubClient(channel);

  ResponseStream<ValidatorStatus> subValidatorStatus(bool emitCurrent) {
    return _client.subValidatorStatus(
        ValidatorStatusSubRequest(emitCurrent: emitCurrent));
  }
}

// -----------------
// ValidatorInfoSubClient
// -----------------
class _ValidatorInfoSubClient {
  final ValidatorInfoSubClient _client;

  _ValidatorInfoSubClient(ClientChannelBase channel)
      : _client = ValidatorInfoSubClient(channel);

  ResponseStream<ValidatorInfo> subValidatorInfo(bool emitCurrent) {
    return _client
        .subValidatorInfo(ValidatorInfoSubRequest(emitCurrent: emitCurrent));
  }
}

// -----------------
// ValidatorStatsSubClient
// -----------------
class _ValidatorStatsSubClient {
  final ValidatorStatsSubClient _client;

  _ValidatorStatsSubClient(ClientChannelBase channel)
      : _client = ValidatorStatsSubClient(channel);

  ResponseStream<ValidatorStats> subValidatorStats() {
    return _client.subValidatorStats(Empty());
  }
}

// -----------------
// Consolidated ValidatorClient
// -----------------
class ValidatorClient {
  final ValidatorOpsClient _validatorOpsClient;
  final _ValidatorStatusSubClient _validatorStatusSubClient;
  final _ValidatorInfoSubClient _validatorInfoSubClient;
  final _ValidatorStatsSubClient _validatorStatsSubClient;

  ValidatorClient(ClientChannelBase? channel)
      : _validatorOpsClient =
            ValidatorOpsClient(channel ?? defaultLuzidChannel),
        _validatorStatusSubClient =
            _ValidatorStatusSubClient(channel ?? defaultLuzidChannel),
        _validatorInfoSubClient =
            _ValidatorInfoSubClient(channel ?? defaultLuzidChannel),
        _validatorStatsSubClient =
            _ValidatorStatsSubClient(channel ?? defaultLuzidChannel);

  Future<ValidatorOpsResponse> validatorOps(ValidatorOpsRequest request) {
    return _validatorOpsClient.validatorOps(request);
  }

  ResponseStream<ValidatorStatus> subValidatorStatus(bool emitCurrent) {
    return _validatorStatusSubClient.subValidatorStatus(emitCurrent);
  }

  ResponseStream<ValidatorInfo> subValidatorInfo(bool emitCurrent) {
    return _validatorInfoSubClient.subValidatorInfo(emitCurrent);
  }

  ResponseStream<ValidatorStats> subValidatorStats() {
    return _validatorStatsSubClient.subValidatorStats();
  }
}
